Test file when ran individually works:
```
➜  owncloud2 git:(bug-fix/wrong-test-behavior-FileTest-testPutWithModifyRun) ✗ make test-php TEST_DATABASE=pgsql TEST_PHP_SUITE=apps/dav/tests/unit/Connector/Sabre/FileTest.php 
PHPUNIT="/home/sujith/test/owncloud2/lib/composer/phpunit/phpunit/phpunit" build/autotest.sh pgsql apps/dav/tests/unit/Connector/Sabre/FileTest.php
Using PHP executable /usr/bin/php
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Parsing all files in lib/public for the presence of @since or @deprecated on each method...

Using database oc_autotest
Setup environment for pgsql testing on local storage ...
Installing ....
Cannot load Xdebug - it was already loaded
ownCloud was successfully installed
Testing with pgsql ...
Cannot load Xdebug - it was already loaded
/home/sujith/test/owncloud2/lib/composer/phpunit/phpunit/phpunit --configuration phpunit-autotest.xml --coverage-clover autotest-clover-pgsql.xml --coverage-html coverage-html-pgsql --log-junit autotest-results-pgsql.xml ../apps/dav/tests/unit/Connector/Sabre/FileTest.php 
Cannot load Xdebug - it was already loaded
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.7-1+ubuntu16.04.1+deb.sury.org+1 with Xdebug 2.6.0
Configuration: /home/sujith/test/owncloud2/tests/phpunit-autotest.xml

................................................................. 65 / 84 ( 77%)
...................                                               84 / 84 (100%)

Time: 39.42 seconds, Memory: 76.00MB

OK (84 tests, 472 assertions)

Generating code coverage report in Clover XML format ...^CMakefile:179: recipe for target 'test-php' failed
make: *** [test-php] Interrupt

➜  owncloud2 git:(bug-fix/wrong-test-behavior-FileTest-testPutWithModifyRun) ✗
```
Debugging further.

--- a/apps/dav/lib/Connector/Sabre/File.php
+++ b/apps/dav/lib/Connector/Sabre/File.php
@@ -223,6 +223,10 @@ class File extends Node implements IFile, IFileNode {
 			$view = \OC\Files\Filesystem::getView();
 			if ($view) {
 				$run = $this->emitPreHooks($exists);
+				if ($run === false) {
+					$view->unlockFile($this->path, ILockingProvider::LOCK_SHARED);
+					return null;
+				}
 			} else {
 				$run = true;
 			}

--- a/apps/dav/lib/Connector/Sabre/File.php
+++ b/apps/dav/lib/Connector/Sabre/File.php
@@ -183,7 +183,7 @@ class File extends Node implements IFile, IFileNode {
 		list($storage, $internalPath) = $this->fileView->resolvePath($this->path);
 		try {
 			$target = $partStorage->fopen($internalPartPath, 'wb');
-			if ($target === false) {
+			if (!\is_resource($target)) {
 				\OCP\Util::writeLog('webdav', '\OC\Files\Filesystem::fopen() failed', \OCP\Util::ERROR);
 				// because we have no clue about the cause we can only throw back a 500/Internal Server Error
 				throw new Exception('Could not write file contents');

--- a/apps/dav/tests/unit/Connector/Sabre/FileTest.php
+++ b/apps/dav/tests/unit/Connector/Sabre/FileTest.php
@@ -296,8 +296,6 @@ class FileTest extends TestCase {
 	 * value in the listener. Once it is modified check the exception returned
 	 * from main function. The reason for exception is put from Sabre/File.php
 	 * throws exception
-	 *
-	 * @expectedException \Sabre\DAV\Exception
 	 */
 	public function testPutWithModifyRun() {
 		$calledUploadAllowed = [];
@@ -312,6 +310,8 @@ class FileTest extends TestCase {
 			->setMethods(['fopen'])
 			->setConstructorArgs([['datadir' => \OC::$server->getTempManager()->getTemporaryFolder()]])
 			->getMock();
+		$storage->method('fopen')
+			->willReturn($this->getStream('qwertz'));
 		Filesystem::mount($storage, [], $this->user . '/');
 		/** @var View | \PHPUnit_Framework_MockObject_MockObject $view */
 		$view = $this->getMockBuilder(View::class)
@@ -337,7 +337,7 @@ class FileTest extends TestCase {
 
 		$file = new File($view, $info);
 
-		$file->put('test data');
+		$file->put($this->getStream('test data'));
 
 		$this->assertInstanceOf(GenericEvent::class, $calledUploadAllowed[1]);
 		$this->assertArrayHasKey('run', $calledUploadAllowed[1]);

--- a/apps/dav/tests/unit/Connector/Sabre/FileTest.php
+++ b/apps/dav/tests/unit/Connector/Sabre/FileTest.php
@@ -293,14 +293,12 @@ class FileTest extends TestCase {
 
 	/**
 	 * Test the run value when put is called. And then try to modify the run
-	 * value in the listener. Once it is modified check the exception returned
-	 * from main function. The reason for exception is put from Sabre/File.php
-	 * throws exception
+	 * value in the listener.
 	 */
 	public function testPutWithModifyRun() {
 		$calledUploadAllowed = [];
-		\OC::$server->getEventDispatcher()->addListener('file.beforeUpdate', function (GenericEvent $event) use (&$calledUploadAllowed) {
-			$calledUploadAllowed[] = 'file.beforeUpdate';
+		\OC::$server->getEventDispatcher()->addListener('file.beforeCreate', function (GenericEvent $event) use (&$calledUploadAllowed) {
+			$calledUploadAllowed[] = 'file.beforeCreate';
 			//Now modify run
 			$event->setArgument('run', false);
 			$calledUploadAllowed[] = $event;
@@ -319,30 +317,17 @@ class FileTest extends TestCase {
 			->setConstructorArgs([])
 			->getMock();
 
-		$view->expects($this->atLeastOnce())
-			->method('resolvePath')
-			->will($this->returnCallback(
-				function ($path) use ($storage) {
-					return [$storage, $path];
-				}
-			));
-
-		$view->expects($this->any())
-			->method('getRelativePath')
-			->will($this->returnArgument(0));
-
-		$info = new FileInfo('/test.txt', $this->getMockStorage(), null, [
-			'permissions' => Constants::PERMISSION_ALL
-		], null);
-
-		$file = new File($view, $info);
-
-		$file->put($this->getStream('test data'));
-
+		$this->assertNull($this->doPut('/test1.txt'));
 		$this->assertInstanceOf(GenericEvent::class, $calledUploadAllowed[1]);
 		$this->assertArrayHasKey('run', $calledUploadAllowed[1]);
 		$this->assertFalse($calledUploadAllowed[1]->getArgument('run'));
-		$this->assertEquals('file.beforeUpdate', $calledUploadAllowed[0]);
+		$this->assertEquals('file.beforeCreate', $calledUploadAllowed[0]);
+
+		//Remove the listener for the event 'file.beforeCreate'
+		$eventListeners = \OC::$server->getEventDispatcher()->getListeners('file.beforeCreate');
+		foreach ($eventListeners as $eventListener) {
+			\OC::$server->getEventDispatcher()->removeListener('file.beforeCreate', $eventListener);
+		}
 	}
 
 	/**
@@ -406,7 +391,7 @@ class FileTest extends TestCase {
 				$calledCreateAllowed[] = 'file.beforeCreate';
 				$calledCreateAllowed[] = $event;
 			});
-			\OC::$server->getEventDispatcher()->addListener('file.beforeCreate', function (GenericEvent $event) use (&$calledWriteAllowed) {
+			\OC::$server->getEventDispatcher()->addListener('file.beforeWrite', function (GenericEvent $event) use (&$calledWriteAllowed) {
 				$calledWriteAllowed[] = 'file.beforeWrite';
 				$calledWriteAllowed[] = $event;
 			});
@@ -840,19 +825,8 @@ class FileTest extends TestCase {
 		);
 
 		// action
-		$thrown = false;
-		try {
-			$this->doPut('/foo.txt');
-		} catch (Exception $e) {
-			$thrown = true;
-		}
+		$this->assertNull($this->doPut('/foo.txt'));
 
-		// objectstore does not use partfiles -> no move after upload -> no exception
-		if ($this->runsWithPrimaryObjectstorage()) {
-			$this->assertFalse($thrown);
-		} else {
-			$this->assertTrue($thrown);
-		}
 		$this->assertEmpty($this->listPartFiles(), 'No stray part files');
 	}
 

